Add component folder guidelines

perrupa opened the pull request:
## What?

We were looking to web-foundation for some guidance on react component structure and found a `TODO`, so I figured I'd take a stab at some baseline rules for how to start structuring components and scaling the folder structure as it grows.

## Why?

My intention here is to set some simple rules for when and how to scale up a component, but also provide some reasoning for why you'd want to create folders. I want to discourage cargo-culting the folder structure of complex components when often we're creating relatively simple components. The intent of the rules are to provide guidance on when to group things without providing too much wiggle room for misinterpretation.

## Notes

Is this repo our new source of truth? Do we have any intention of keeping [`shopify/javascript`](https://github.com/shopify/javascript) up to date or getting docs from this repo published to https://development.shopify.io/?

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+.
+└── components
+    ├── ...
+    ├── MyComponent.js
```
These should default to `typescript`, so `.tsx` here.

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+
+#### 3. As your component grows, group like-files in folders.
+
+Common types to group inside a component:
```
Can we represent `hooks` here as well by way of a `hooks` directory or `hooks.ts` file. You can link to the section below for more details.

perrupa replied:
Great idea 👍

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+    └── MyComponent
+        ├── MyComponent.js
+        ├── MyComponent.scss
+        └── MyComponent.test.js
```
This should be tucked in a `tests` folder 100% of the time.

perrupa replied:
I would challenge this. Most components I see with a `tests/` folder (in partners at least) have only one test file in them.

I understand that when you have lots of different tests there is a benefit in grouping them in a `tests/` folder (I deliberately used that as my first example 😄). What I would like to see less of is single-file-containing folders, I don't see a benefit of isolating a single test in a folder. Which is why I prefer it as optional, and mention it when grouping the files.

I'm totally open to hearing the reasoning though if you've got the time 🤗

cartogram replied:
Basically should a single test file go inside a folder. I think it's fair to argue against that 👍 I am likely coming at this with a `web`-bias [where](https://github.com/Shopify/web/tree/master/app/foundation/App/tests) [we](https://github.com/Shopify/web/tree/master/app/foundation/App/components/AppDetailsProvider/tests) [always](https://github.com/Shopify/web/tree/master/app/foundation/App/components/Monorail/tests)  [have](https://github.com/Shopify/web/tree/master/app/foundation/App/components/ContentSecurityPolicy/tests)  a `tests` folder, so seeing them run loose with the other files looks strange.

perrupa replied:
Yeah, those are good examples of why I'm pushing back on that convention 😄 

e.g. 
https://github.com/Shopify/web/tree/master/app/foundation/App/components/ContentSecurityPolicy

```
└── components
    └── ContentSecurityPolicy
        ├── ContentSecurityPolicy.tsx
        └── tests/
```
reads just as clean as
```
└── components
    └── ContentSecurityPolicy
        ├── ContentSecurityPolicy.tsx
        └── ContentSecurityPolicy.test.tsx
```
with the exception that in the 2nd example I can see and open the test, and as a bonus I can see there's only one 😉 

But this is good to know 👌

cartogram commented on Best practices/React/Organization.md:
````diff
@@ -4,11 +4,109 @@
+or
+
+```
+components
````
Between the 2, the former is preferred so I would just remove this alternate example.

perrupa replied:
I think they're both valid, I also chose to leave out these two beasts

```

components
└── ImportInventory
    ├── index.ts
    ├── ImportInventory.tsx
    ├── components
    │   ├── ImportCreated
    │   │   ├── ImportCreated.tsx
    │   │   └── tests
    │   │      └── ImportCreated.test.tsx
    │   ├── ImportSubmitted
    │   │   ├── ImportSubmitted.tsx
    │   │   └── tests
    │   │      └── ImportSubmitted.test.tsx
    │   └── InitialState
    │       ├── InitialState.tsx
    │       └── tests
    │           └── InitialState.test.tsx
    ├── graphql
    │   ├── InventoryImportCreateMutation.graphql
    │   ├── InventoryImportSubmitMutation.graphql
    │   └── InventoryStagedUploadMutation.graphql
    └── tests
        └── ImportInventory.test.tsx
```

or 

```

components
└── ImportInventory
    ├── index.ts
    ├── ImportInventory.tsx
    ├── components
    │   ├── ImportCreated
    │   │   ├── ImportCreated.tsx
    │   │   ├── graphql
    │   │   │   └── InventoryImportCreateMutation.graphql
    │   │   └── tests
    │   │      └── ImportCreated.test.tsx
    │   ├── ImportSubmitted
    │   │   ├── ImportSubmitted.tsx
    │   │   ├── graphql
    │   │   │   └── InventoryImportSubmitMutation.graphql
    │   │   └── tests
    │   │      └── ImportSubmitted.test.tsx
    │   └── InitialState
    │       ├── InitialState.tsx
    │       ├── graphql
    │       │   └── InventoryStagedUploadMutation.graphql
    │       └── tests
    │           └── InitialState.test.tsx
    └── tests
        └── ImportInventory.test.tsx
```

If this file structure isn't a great example though, let's come up with a better one. I can't remember where specifically but I think this was from `web` (I also trimmed out a shit ton of index.js files that were just all over)

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+- Multiple tests into `./tests/`
+- graphQL queries and mutations in `./graphql/`
+- Subcomponents of the component in it's own `./components/` subfolder
+- utility files containing helpers in `./utils/`
```
```suggestion
- Utility files containing helpers in `./utilities/`
```

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+## Application
 
-### `/app/components` and `<Component>/components`
+### React Components and Sub-Components
```
```suggestion
### React Components and Sub-components
```

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+Common types to group inside a component:
+- Multiple tests into `./tests/`
+- graphQL queries and mutations in `./graphql/`
+- Subcomponents of the component in it's own `./components/` subfolder
```
```suggestion
- Sub-components of the component in it's own `./components/` subfolder
```

cartogram commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,109 @@
+
+Common types to group inside a component:
+- Multiple tests into `./tests/`
+- graphQL queries and mutations in `./graphql/`
```
```suggestion
- GraphQL queries and mutations in `./graphql/`
```

emarchak commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,111 @@
-### `/app/components` and `<Component>/components`
+### React Components and Sub-components
+
+React Components should always exist in a `components/` folder, but the component's internal structure may change depending on their complexity.
```
Could we define the difference between the three, `foundation`,`features`, and `components`?

michenly replied:
`root/foundation`: the base of the application that is usually render once. Such as Providers of the application, routing setup, and frame (sidebar, topbar) of the application. It typically contains react component that is not meant for reuse anywhere else.

`root/features`: different parts of the application, typically corresponded to top level url. Majority of the application lives here.

`root/components`: top level components works much like the rest of the `components`. It contains React components meant to be share and re-use by any file in the same folder or below.

perrupa replied:
I feel like that might be project level config and I'm trying to scope this exclusively to components,  but that's a great point. Are `root/features` similar to `sections`? I've actually never heard these conventions before, are they unique to `web`, or do we use them elsewhere as well?

Maybe I should just remove this sentence? or change `always` to usually and list the exceptions you mentioned?

michenly replied:
`root/features` = `sections`, just the future version :)

I do agree it may be out of scope. Deleting the word `always` and adding the `root/components` explanation in here to the extend it make sense should be enough.

> React Components can often be found in `components` folder. These folder are meant to house share and re-useable React Components that should only be consume by other components in the same folder or below.

emarchak commented on Best practices/React/Organization.md:
````diff
@@ -4,11 +4,111 @@
+import {MyComponent} from './components/MyComponent/MyComponent';
+```
+
+You have several options to fix this problem:
````
The second option allows for any kind of internal structure, and easy imports and refactors. The rest of the options don't seem as intuitive to me. 

> create an `index.tsx` inside the component's folder that exports `MyComponent.tsx`

donnielrt commented on Best practices/React/Organization.md:
```diff
@@ -4,11 +4,111 @@
+
+#### 1. Start with a file and a test
+
+At its simplest a component will exist as a component and a test at the `components/` level.
```
Should we recommend starting with an `index.ts` that exports the component right off the bat?

TheMallen commented:
@perrupa do you have any plans to come back to this? I can pick it up if you don't

perrupa commented:
@TheMallen and I paired on this a few days ago and simplified things a whole lot. Removed a lot of decisions in favor of a minimal set of easy-to-apply rules.

ismail-syed commented on handbook/Best practices/React/Organization.md:
```diff
@@ -4,11 +4,62 @@
-## `/app`
+## Application
 
-### `/app/components` and `<Component>/components`
```
Can we still leave these top-level folder paths in? I think there useful to have.

TheMallen commented:
@perrupa seems like this should be good to merge now?
